\subsection{Human data results}\label{app:human}

\input{\dir/tab-human-data}

\begin{figure}[H]
  \centering
  \includegraphics[scale=0.55]{plots/real_full_without_bio_var.pdf}\\
  \includegraphics[scale=0.55]{plots/real_full_with_bio_var.pdf}%
  \caption[Runtime on long reads of human data]{\textbf{Runtime on long reads of
    human data} (logarithmic, $\geq 500\kbp$). Each dot corresponds to an
    alignment of a sequence pair. The ONT reads are without (top) and with
    (bottom) genetic variation. Runtime is capped at $\qty{100}{s}$.}
  \label{fig:human-full}
\end{figure}

Statistics on our human data are presented in~\cref{tab:hg}. We compare all our
heuristics in~\cref{fig:human-full}.

For ONT reads without genetic variation, \SH is faster than other methods in
median, but slower for sequences with high divergence and larger gaps.
\CSH is slightly slower than \SH due to additional bookkeeping without
significant benefit for the heuristic. Penalizing indels with \GCH improves
performance several times.

When genetic variation is included, \SH and \CSH do not sufficiently penalize
long gaps, and \A with \GCH estimates the remaining edit distance significantly
better. The diagonal transition optimization considerably speeds up the
alignment of long indels where the search regresses to quadratic.
