Sequence alignment is the task of finding similarities between sequences. It has
been a central building block in molecular biology since DNA, RNA and protein
sequences were first obtained half a century ago. Sequence alignment is applied
in research, and medicine, with applications such as evolutionary biology,
genome assembly, oncology and many others. The correctness of the analyses of
the growing amounts of genetic data require algorithms with high accuracy and
speed. Moreover, the ongoing transition from using single genome references to
using pangenomes motivates adaptation of the classic alignment algorithms.

We consider two types of pairwise alignment problems: semi-global mapping of a
set of DNA reads to a general graph reference representing a pangenome, and
global (or end-to-end) alignment of two sequences. Modern optimal algorithms
perform uninformed search -- they neglect information from the unaligned parts of
the sequences. As a result, their runtime and memory footprint increase
near-quadratically in the sequence length even if the sequences are related.
Despite the negative theoretical results affirming that it is unlikely that
strongly subquadratic algorithms exist for the worst case, in practice, our
approach overcomes this quadratic limitation by exploiting information from the
whole length of related sequences. We follow a principled approach that
guarantees to find a best alignment as equating it to finding a shortest path in
an alignment graph.

\A is a classic shortest path algorithm, which leverages heuristics to direct
the search, and yet it finds provably shortest paths given an admissible
(optimistic) heuristic function which estimates the remaining distance
(corresponding to the cost to align the remaining suffixes of the sequences).
Curiously, despite previous attempts to apply \A to multiple sequence alignment
(MSA), it has not been adopted even more pairwise sequence alignment. In the
present thesis we investigate how to do that. We (i) phrase the alignment
problem as a shortest path problem, (ii) suggest a novel highly-informed
admissible heuristic, (iii) design efficient algorithms and data structures for
computing the heuristic, (iv) prove their optimality, (v) implement the
presented algorithms, and (vi) compare their performance and scaling to other
optimal algorithms. Our approach is provably optimal according to edit distance,
its runtime empirically scales subquadratically (and sometimes even
near-linearly) with the output size, which translates to orders of magnitude of
speedup compared to state-of-the-art optimal algorithms.

Throughout this thesis, we demonstrate how to encompass various dimensions of
the input complexity while preserving the speed and optimality. First, we
demonstrate how to apply a trie index to scale the mapping runtime sublinearly
with the reference size. Then, we introduce a novel seed heuristic for \A which
enables aligning of long sequences (up to millions of base pairs) near-linearly
with their length. Last, we extend the seed heuristic with inexact matching and
match chaining to raise the tolerated error rate (up to 30\%). Interestingly,
our seed usage challenges the omnipresent seed-(chain)-extend paradigm to
sequence alignment.

These first steps to scalable optimal alignment using \A give rise to a number
of research directions: approaching other alignment types, more general
optimization metrics, relaxed optimality guarantees, performance analyses,
improved heuristics, applications outside of biology, and more performant
algorithms and implementations. Perhaps, one day the \A approach to alignment
will mature enough to match the performance of suboptimal algorithms for both
linear and pangenome applications.