%*******************************************************
% Abstract
%*******************************************************
%\renewcommand{\abstractname}{Abstract}
\pdfbookmark[1]{Abstract}{Abstract}
\begingroup
\let\clearpage\relax
\let\cleardoublepage\relax
\let\cleardoublepage\relax

\chapter*{Abstract}

% Context
Sequence alignment is a class of problems aiming at unrevealing the relationship
between sequences. These problems are of central interest in computational
molecular biology for more than half a century and they are routinely applied in
evolutional biology, genome assembly, read mapping, variation detection and
others.

% Contribution overview
Alignment algorithms involve a common tradeoff between optimality guarantees and
performance. In this thesis, we consider a shortest path formulation of
alignment and explore the application of the A* algorithm. Most importantly,
this approach enables us to design heuristic functions which combine optimality
guarantees and polynomial speed ups.

% Problem statement
Depending on the the number of aligned sequences, there is pairwise alignment
and multiple sequence alignment (MSA). Depending on the parts of the sequences
that are aligned to each other, we differentiate global, local and various
semi-global alignemnts. There are generalizations to sequence-to-sequence
alignment, including aligning to nonlinear structures, such as directed acyclic
graphs, DAGs, general graphs and others. These structures are nowadays becomming
more present as a compressed form of representing a set of references to which a
sequence can be aligned. Often, one best alignment is sufficient but finding
several best (top-K) alignments.

% Accuracy and Metrics
The number of possible alignments grow exponentially with length. The usual
underlying question to finding ``correct'' alignments. Regarding the precision
of alignment, one is usually interested in base-to-base (aka letter-to-letter)
correspondence between the sequences, even though for some applications a less
detailed solution is sufficient: only the similarity between sequences or the
location where a read maps to a reference. Exact alignment is only useful for
very short sequences (often kmers), and for all other cases the optimized metric
may be hamming distance, edit distance (unit costs), Levenshtein distance,
affine costs, convex and concave costs, general costs and others. 

% Algorithms and Asymptotics
The classical dynamic solutions for are the so called Needleman-Wunsch.
Currently, researchers still face the In the context of read mapping, a set of
reads is aligned to the same reference sequence so an indexing procedure is
often useful for the performance.

% Data variation
A good algorithm should be optimal, complete, performant (fast and low on
memory) on short and long sequences, various error rates and error models,
and reference complexity (topology and enthropy).

% Goals
%Accuracy (approximate vs probabilistic vs exact/optimal algorithm), Runtime,
%memory usage.

% A*

% Tools and Performance
Many tools do not have a well-stated problem they optimize.

% Future work
Probabilistic approach.

\endgroup

\cleardoublepage%

\begingroup
\let\clearpage\relax
\let\cleardoublepage\relax
\let\cleardoublepage\relax

\begin{otherlanguage}{ngerman}
\pdfbookmark[1]{Zusammenfassung}{Zusammenfassung}
\chapter*{Zusammenfassung}

Deutsche Zusammenfassung hier.

\end{otherlanguage}

\endgroup

\vfill