\graphicspath{{\dir/}}

\chapter{Introduction} \label{ch:introduction}

%\dictum{%
%   Provably optimal and heuristically fast.}
%\vskip 1em

Sequence alignment is a class of combinatorial problems that is of primary
importance for analysis of genetic data. Algorithms and tools for alignment have
been thoroughly developed and routinely used for genome assembly, RNA
quantification, detecting splicing, oncology, multiple sequence alignment (MSA),
and evolutionary biology. Types of sequence alignment include global alignment,
semi-global alignment, mapping, local alignment, and others. For each type, a
common tradeoff that had to be done is between the alignment accuracy and the
performance to find it.

% graph reference
An additional difficulty is the fact that in the
upcoming pangenomic era, these algorithms must be also applicable to complex
graph structures. For more than 60 years, a linear sequence has been extremely
useful as a representation of a single genome. The affordability of sequencing
technologies enables not only to sequence genomes deeper but also to sequence
many genomes (e.g. of organisms or single cells), building a pangenome (an
abstracted genome that represents the genetic variation of a whole clade). The
shift towards population studies in the last decade motivates the adoption of
graph data structures which serve as compressed representations of collections
of related genomes (genomes are paths in the graph).

% A* context 
An optimal alignment can naturally be represented as a shortest path in an
alignment graph (equivalent to the DP table). In order to find such a shortest
path with minimal exploration, we instantiate the A* algorithm with a novel
heuristic function based on the unaligned parts of the sequences. This
additional information is a problem-specific heuristic function and it heavily
determines the efficiency of the search. For any explored state by A*, this
heuristic function should compute a lower bound on the remaining path length, or
more specifically, the minimal cost of edit operations needed to align the
remaining sequences.

\paragraph{Main contributions}
We have effectively applied the \A algorithm to optimal sequence alignment. We
demontrated that the additional information from the whole sequence can improve
the scaling with query length, reference size and error rate, substantially
decrease the necessary computations, and result in algorithms that are orders of
magnitude faster than existing optimal algorithms. We apply the \A approach to
two types of alignment: semi-global (mapping) and global.

%only optimal alignment in this thesis
%asymptotics analysis
%????? linear I/O but quadratic optimal

\input{\dir/contributions}
\input{\dir/problem}
\input{\dir/prelims}
