\section{Previous work}

It may come as a surprise that another approach to such a basic problem as
alignment appears 60 years after the problem was first efficiently solved. Here
we speculate about the possible reasons for it: focus on new technology and
data, not believing that optimal solutions could be efficient.

%combining several existing ideas new (\A, trie, seeds)

Most of the techniques this thesis builds upon have been known for many decades
and have also been heavily motivated by applications in molecular biology. The
global alignment problem exists since
\citeyear{vintsyuk1968speech}~\cite{vintsyuk1968speech,needleman1970general}.

The problem of approximate/fuzzy string search (semi-global alignment of one
query) has been efficiently approached a decade later, in
\citeyear{sellers1980theory}~\cite{sellers1980theory,smith1981identification}.
more recent and additionally to the issues with handling errors, a new dimension
of complexity appears by the need to locate the position in the reference. The
problem of searching for multiple sequencing has been approached
since\citeyear{pearson1988improved}~\cite{pearson1988improved} and became
central to read alignment of high-throughput sequencing. The specifics with
semi-global alignment requires a trie-like index which is well-used in the
fiels, and known since
\citeyear{thue1912gegenseitige}~\cite{thue1912gegenseitige} and used in
informatics since~\citeyear{de1959file}~\cite{de1959file}.

An important factor for the development of the sequencing algorithms is the
technological advancement~\cite{alser2021technology}. One possible reason is the
combination of massive new data and advancement in computation hardware allowing
for various kinds of parallelization. Thus the advancements with bit-parallel
algorithms (CITE Myears, Mikko).

A major importance for our apporach is the usage of seeds to build good
heuristic functions that drive the \A search. Seeds are cousins of kmers which
are popular in sequence alignment since de Bruijn Graphs were applied for genome
assembly. Further kmers are cousins of ngrams which are popular in computational
linguistics since .

% parallelization
% sketches

% Optimal DP-based approaches
\paragraph{Optimal Alignment}
Current optimal alignment algorithms reach the impractical $\Oh(nm)$ runtime
that has been shown to be a lower bound for the worst-case edit distance
computation~\cite{backurs2015edit}. In this light, approaches for improving the
efficiency of optimal alignment have taken advantage of specialized features of
modern CPUs to improve the practical runtime of the Smith-Waterman dynamic
programming (DP) algorithm~\cite{smith_comparison_1981} considering all possible
starting nodes. These use modern SIMD instructions (\eg
\vg~\cite{garrison_variation_2018} and \pasgal~\cite{jain_accelerating_2019}) or
reformulations of edit distance computation to allow for bit-parallel
computations in \graphaligner \footnote{We refer as \bitparallel to to the
bit-parallel DP algorithm implemented in \graphaligner tool
\cite{rautiainen_bitparallel_2019}.}~\cite{rautiainen_bitparallel_2019}. Many of
these, however, are designed only for specific types of genome graphs, such as
{\itshape de Bruijn}
graphs~\cite{liu_debga_2016,limasset2019toward} and
variation graphs~\cite{garrison_variation_2018}. A compromise often made when
aligning sequences to cyclic graphs using algorithms reliant on directed acyclic
graphs involves the computationally expensive ``DAG-ification'' of graph
regions~\cite{kavya_sequence_2019,garrison_variation_2018}.