\section*{Related work}
\addcontentsline{toc}{section}{\protect\numberline{}{Related work}}

Over the last half-century, many algorithms have been developed for sequence
alignment~\citep{navarro2001guided}. With the increasing quantity and length of
genomic sequences, the need for fast and accurate algorithms is steadily
increasing~\cite{alser2021technology}. There are multiple directions in which
the sequence alignment algorithms are being developed: scalable algorithms,
generalizing the optimization metric, employing heuristics for approximate
solutions, adapting existing algorithms to novel types of data, exploiting
parallel computing, implementing fast and memory-efficient software. This thesis
focuses on algorithmic approaches to sequence alignment, so here we outline the
leading existing optimal algorithms, the suboptimal seeding heuristics that we
use in an optimal way, and the \A shortest path algorithm, which will be the
base for our provably optimal and heuristically fast approach.

\paragraph{Dynamic programming for global alignment}
The first efficient solution for global alignment, known as the Needleman-Wunsch
algorithm, got developed around~\citeyear{vintsyuk1968speech}~
\cite{vintsyuk1968speech,needleman1970general}. This algorithm is based on the
dynamic programming~(DP) technique~\cite{bellman1954theory} of splitting a task
into overlapping subtasks~(or \emph{states}) each of which can be solved once
and then reused. Each node of the alignment graph (aligning a prefix of the
first sequence to a prefix of the second sequence) is a state in the DP. Each
subtask can be solved by reducing it to already solved alignments of shorter
prefixes and extending them by a last operation of matching or mismatching the
last letters from the prefixes, inserting one or inserting the other
(equivalent to deleting from the first sequence). Each subtask is solved for
$\Oh(1)$ so the quadratic number of prefix pairs is solved for quadratic overall
time, even though the number of possible alignments is exponential.
Independently, it was applied to compute edit distance for biological
sequences~\citep{needleman1970general,sankoff1972matching,sellers1974theory,wagner1974string}.
This well-known algorithm is implemented in modern aligners like
\seqan~\citep{reinert2017seqan} and \parasail~\citep{daily2016parasail}.
Improving the performance of these quadratic algorithms has been a central goal
in later works.

\paragraph{Dynamic programming for semi-global alignment}
Semi-global alignment is a related problem to global alignment, but in addition
to the need to align letter-to-letter, it also requires determining a starting
position in the reference. This task has also been efficiently solved using a
similar DP approach~\cite{sellers1980theory,smith1981identification} taking
$\Oh(nm)$ to align a single query of length $n$ to a reference of length $m$. It
is highly redundant to explore the whole reference for each alignment, so
various data structures were suggested that can be precomputed to index the
reference, and then reuse this index to quickly align multiple query sequences
~\citeyear{pearson1988improved}~\cite{pearson1988improved} and has become
central to read alignment of high-throughput sequencing.

Current optimal alignment algorithms reach the impractical $\Oh(nm)$ runtime
that is a lower bound for the worst-case edit distance
computation~\cite{backurs2015edit}. In this light, approaches for improving the
efficiency of optimal alignment have taken advantage of specialized features of
modern CPUs to improve the practical runtime of the Smith-Waterman dynamic
programming (DP) algorithm~\cite{smith_comparison_1981} considering all possible
starting nodes.

\paragraph{Suboptimal heuristic algorithms}
All optimal solutions are near-quadratic and this is justified by the
theoretical lower bound. To speed up beyond this near-quadratic barrier, current
practical algorithms break the optimality guarantee with the hope that the
produced alignments are accurate enough or give up on alignment altogether.
Heuristics are employed for both sequence-to-sequence alignment and
sequence-to-graph alignment, to keep the computation
tractable~\cite{altschul_basic_1990,langmead_fast_2012,garrison_variation_2018},
especially for large populations of human-sized genomes. In the last decades,
approximate and alignment-free methods satisfied the demand for faster
algorithms that process huge volumes of genomic
data~\cite{kucherov2019evolution}. While heuristics may produce acceptable
alignments in many cases, it fundamentally does not provide quality guarantees,
resulting in suboptimal alignment accuracy.

\paragraph{Seed-and-extend}
Since optimal alignment is often intractable, many aligners use heuristics, most
commonly the \emph{seed-and-extend}
paradigm~\cite{altschul_basic_1990,langmead_fast_2012,li_fast_2009}. In this
approach, alignment initiation sites (\emph{seeds}) are determined, which are
then \emph{extended} to form the \emph{alignments} of the query sequence. The
fundamental issue with this approach, however, is that the seeding and extension
phases are mostly decoupled during alignment. Thus, an algorithm with a provably
optimal extension phase may not result in optimal alignments due to the
selection of a suboptimal seed in the first phase. In cases of high sequence
variability, the seeding phase may even fail to find an appropriate seed from
which to extend.

\paragraph{Shortest path formulation and \A for sequence alignment}
\A is a shortest path algorithm that generalizes \dijkstra by directing the
search towards the target~\citep{hart1968formal}. \A is an \emph{informed}
search algorithm -- it relies on a problem-specific heuristic function that
estimates the remaining distance from each explored node to the target. If this
heuristic is \emph{admissible}, i.e. it never overestimates the actual distance
to the target, \A is guaranteed to find an optimal path. If the heuristic is
accurate -- similar to the actual remaining distance, then \A finds the path
with minimal exploration. If the heuristic can be computed efficiently, the
overall runtime stays low. Designing a heuristic function with all of these
properties has been challenging in the context of sequence alignment. Attempts
to solve optimal pairwise alignment by \A were not scalable due to overly
simplistic heuristic functions~\cite{dox2018efficient}. Tries to solve multiple
sequence alignment~(MSA) by \A were also impractical due to the persisting
exponential scaling with the number of
sequences~\citep{lermen2000practical,zhou2002multiple,mcnaughton2002memory}.

\paragraph{Optimal semi-global alignment to graph references}
First attempts to include variation into the reference data structure were made
by augmenting the local alignment method to consider alternative walks during
the extend step~\cite{schneeberger_simultaneous_2009,palmapper}. This approach
has since been extended from the linear reference case to graph references. To
represent the non-reference variation of multiple references during the seeding
stage, HISAT2 uses generalized compressed suffix
arrays~\cite{siren_indexing_2014} to index walks in an augmented reference
sequence, forming a local genome graph~\cite{kim_graphbased_2019}.
VG~\cite{garrison_variation_2018} uses a similar
technique~\cite{siren_indexing_2017} to index variation graphs representing a
population of references. Modern sequence-to-graph aligners use SIMD
instructions (\vg~\cite{garrison_variation_2018} and
\pasgal~\cite{jain_accelerating_2019}) or reformulations of edit distance
computation to allow for bit-parallel computations in
\graphaligner~\cite{rautiainen_bitparallel_2019}. Many of these, however, are
designed only for specific types of genome graphs, such as {\itshape de Bruijn}
graphs~\cite{liu_debga_2016,limasset2019toward} and variation
graphs~\cite{garrison_variation_2018}. A compromise is often made when aligning
sequences to cyclic graphs using algorithms reliant on directed acyclic graphs
involves the computationally expensive ``DAG-ification'' of graph
regions~\cite{kavya_sequence_2019,garrison_variation_2018}.

\paragraph{Trie index}
The specifics with semi-global alignment require a trie-like index which is
well-used in the fields, and known since
\citeyear{thue1912gegenseitige}~\cite{thue1912gegenseitige} and used in
informatics since~\citeyear{de1959file}~\cite{de1959file}.