%\section{Generalizations}
%\todo{Describe mathematically, implement and evaluate}
%
%\subsection{Affine costs}
%
%The state space can be multiplied by the two-element set \{gap opened, gap
%closed\} thus leading to the classical DP problem statement for affine gap costs.
%
%Note that negative costs are not representable because of the triangle
%inequality that both Dijkstra and \A~algorithms rely on for their optimality.
%
%\subsection{Pair reads}
%
%The state space for pair reads is the cross product of the spaces for both
%reads. Every step is made either in the first or the second subspace. The two
%heuristic functions are used (one per read in the pair). in case the two
%substates are too far apart, the cost function is set to $+\infty$, or otherwise
%to the sum of both cost subfunctions. More suffisticated variants of a cost
%function may account more smoothly for the genetic difference between both
%reads.

%\subsection{Phred values}
%
%$$P_{error}(Q) := 10^{-Q/10}$$
%
%We approach the more general task of aligning a read with phred values. To
%account for the phred values, we indivitually modify the costs for matches and
%substitutions depending on the aligned read nucleotide (the cost of insertions
%and deletions does not change). Introducing phred values implies calibrating the
%costs to real probabilities as opposed to abstract relative costs. Thus, we
%distinguish two consequitive tasks: first define the edit costs (by scaling
%the given costs accordingly and applying individual phred corrections), then
%align the read sequence (without phreds) using the modified costs.
%
%\para{Calibrating edit costs}
%\todo{work out the new costs}
%Accounting for phred values in the alignment cost makes matches more expensive
%and substitutions less expensive. This means that even a perfectly matched read will have a cost
%proportional to its length.

%$$\cmatch^\mli{new} \ge \cmatch$$
%$$\csubst^\mli{new} \le \csubst$$

%\para{Aligning with individual edit costs}
%Assume individual edit costs have been already defined to accomodate the
%information from phred values. In order to preserve the \A~correctness, the
%heuristic function must stay optimistic. In case of a matching seed, the exact
%cost for each match and substitution can be computed. For the lack of matching
%seeds, though, an optimistic estimate is
%$$(m-(\mli{maxErrors}+1))*\cmatch + (\mli{maxErrors}+1)*\min(\csubst, \cins,
%\cdel)$$ where $m$ is the read length.

%\subsection{Spliced reads}
%
%Jump edges from every state $\st{u}{i}$ to state $\st{\mli{supersource}}{i}$ can
%stand for an intron and be attributed relatively high costs.
